{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyNxdbtmwPnWHjDi5nSDtjv+"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#Assignment Questions\n","1. What is the difference between interpreted and compiled languages?\n"," - Compiled languages- Compiled languages are converted directly into machine code that the computer's processor can execute. This conversion happens before the program is run, using a compiler.\n"," - The compiled code runs directly on the hardware, making it generally faster.\n"," - Ex.- Java, c,c++\n"," -Interpreted languages- Interpreted languages are executed line by line by an interpreter. The interpreter reads each line of code and translates it into machine code on the fly.\n"," - ince each line needs to be interpreted as the program runs, interpreted languages are generally slower than compiled languages.\n"," - Ex.- Python, JavaScript\n","\n","2. What is exception handling in Python?\n"," - In Python, exception handling is a mechanism for gracefully handling errors that occur during program execution. This prevents your program from crashing and allows you to control how errors are handled.\n"," - Exceptions are events that disrupt the normal flow of a program. They represent errors or unexpected conditions.\n"," - Ex.- TypeError,FileNotFoundError\n","\n","3. What is the purpose of the finally block in exception handling?\n"," - The finally block is a crucial part of exception handling in Python. It's designed to ensure that certain code is always executed, regardless of whether an exception occurred or not.\n"," - The finally block is also executed even if the control flow within the try block is altered by statements like return, break, or continue. This ensures that cleanup or state restoration operations are performed regardless of how the try block exits.\n","\n","4. What is logging in Python?\n"," - Logging allows you to record events, messages, and data about your program's execution. It helps you track what's happening, identify errors, and debug issues.\n"," - Logging provides a record of events during program execution, helping you monitor its behavior and identify potential problems.\n"," - When errors occur, logging can capture details like the error type, location in the code, and relevant variables, making it easier to diagnose and fix them.\n"," - Logs can be used to audit program activity, analyze usage patterns, and gain insights into how your program is being used.\n","\n","5. What is the significance of the __del__ method in Python?\n"," - The primary significance of the __del__ method is to provide a way to release resources held by an object before it's destroyed. This is crucial for ensuring that resources like files, network connections, or database handles are properly closed and released to prevent leaks or unexpected behavior.\n"," - The __del__ method can be used to perform any finalization tasks associated with an object before it's removed from memory. This might include logging information, updating external systems, or performing any other actions necessary for a clean shutdown of the object.\n"," - The __del__ method is automatically called by Python's garbage collector when an object is no longer referenced by any part of the program. However, the exact timing of garbage collection is not guaranteed, and it might be delayed or even skipped in certain situations.\n","\n","6.  What is the difference between import and from ... import in Python?\n"," - Both statements are used to bring external modules or specific components of modules into your current Python script or environment. However, they differ in how they make those elements accessible.\n"," - import-  Imports the entire module and makes it accessible using its name.\n"," - We need to use the module name to access its functions, classes, or variables.\n"," - from ...- Imports specific components (functions, classes, variables) from a module and makes them directly accessible.\n"," - We can use the component names directly without the module name.\n","\n","7.  How can you handle multiple exceptions in Python?\n"," - There are several ways to handle multiple exceptions in Python\n"," - 1). Using a tuple of exceptions in a single except block - This approach is useful when you want to handle a specific set of exceptions in the same way. The except block will catch any of the exceptions listed in the tuple\n"," - 2). Using multiple except blocks- his approach allows you to handle different exceptions with specific code blocks. The except blocks are evaluated in order, and the first matching exception is handled.\n"," - 3). Using Exception to catch all exceptions- his approach is a catch-all for any exception that might occur. It's useful for general error handling, but it's generally recommended to be more specific with your exception handling when possible.\n","\n","8. What is the purpose of the with statement when handling files in Python?\n"," - The with statement is used to ensure that file resources are properly managed.\n"," - The primary purpose is to automatically close the file after you're done using it.\n"," - The with statement also handles potential exceptions that might occur while working with the file. If an error happens, it ensures the file is closed properly before the exception is propagated.\n"," - It provides a more concise and readable way to work with files compared to manually opening and closing them using try, except, and finally blocks.\n","\n","9. What is the difference between multithreading and multiprocessing?\n"," - Multithreading- It involves creating multiple threads within a single process. These threads share the same memory space and resources of the process.\n"," - Creating threads is generally faster and consumes fewer resources compared to processes.\n"," - Threads can easily share data and communicate with each other.\n"," - Multiprocessing- It involves creating multiple processes, each with its own memory space and interpreter. These processes run independently and can execute code in parallel.\n"," -  Processes can bypass the GIL and utilize multiple CPU cores for CPU-bound tasks, achieving true parallelism.\n"," - Processes are isolated from each other, reducing the risk of shared memory issues and improving stability.\n","\n","10. What are the advantages of using logging in a program?\n"," - Logging provides a detailed record of events, including timestamps, function calls, and variable values, making it easier to identify the source of errors and bugs.\n"," -  Logs can be analyzed after a program has crashed or encountered an issue, providing valuable insights into the sequence of events leading up to the problem.\n"," - Logging helps track the behavior of your program in real-time, allowing you to monitor performance, resource usage, and other key metrics.\n"," - Logging allows you to categorize messages based on their severity (e.g., DEBUG, INFO, WARNING, ERROR, CRITICAL), enabling you to filter and focus on specific types of events.\n"," -  Using a logging framework provides a centralized way to manage logging throughout your program, making it easier to modify and update logging behavior.\n","\n","11. What is memory management in Python?\n"," - Python utilizes automatic memory management, primarily through a mechanism called garbage collection. This means you, as the programmer, generally don't need to manually allocate or deallocate memory for objects. Python handles it behind the scenes.\n"," - Python objects and data structures are stored in a private heap space. This heap is managed internally by the Python memory manager.\n"," - Python keeps track of how many references (or pointers) exist to each object. When an object's reference count drops to zero, it means the object is no longer accessible and can be garbage collected.\n"," - The garbage collector periodically scans the heap for objects with a reference count of zero. These objects are then deallocated, freeing up memory for reuse.\n"," - Python also has a mechanism to detect and collect objects involved in cyclic references.\n","\n","12. What are the basic steps involved in exception handling in Python?\n"," - 1).try Block- Enclose the code that might raise an exception within a try block.\n"," - 2)except Block- Define one or more except blocks to handle specific types of exceptions.\n"," - 3). Optional else Block-  Include an else block to execute code if no exceptions occur in the try block.\n"," - 4). Optional finally Block- Include a finally block to execute code regardless of whether an exception occurred or not. This is often used for cleanup tasks like closing files or releasing resources.\n","\n","13. Why is memory management important in Python?\n"," - If objects are not properly deallocated when they are no longer needed, they can consume memory unnecessarily, leading to memory leaks.\n"," - Memory leaks can significantly impact performance, causing your program to slow down or even crash.\n"," -  Effective memory management ensures that memory is allocated and deallocated efficiently, minimizing waste and maximizing the available resources.\n"," - Improper memory management can lead to crashes or unexpected behavior in your programs.\n"," - By handling memory properly, you can make your programs more robust and less prone to errors.\n","\n","14. What is the role of try and except in exception handling?\n"," - try block- The try block is used to enclose the code that might raise an exception. It essentially defines the \"protected\" section of your code where you anticipate potential errors.\n"," - By placing code within a try block, you signal to Python that you want to handle any exceptions that might occur within that block in a controlled manner.\n"," - except block- The except block is used to define how to handle specific types of exceptions that might be raised within the preceding try block. It acts as the \"handler\" for exceptions.\n"," - When an exception occurs within the try block, Python looks for a matching except block based on the type of exception. If a matching except block is found, the code within that block is executed to handle the exception.\n","\n","15. How does Python's garbage collection system work?\n"," - Python's garbage collection (GC) system is responsible for automatically managing memory, reclaiming it when objects are no longer in use. This helps developers avoid manual memory management and reduces memory leaks. Python uses a combination of reference counting and a cyclic garbage collector to manage memory.\n"," - reference counting- Python primarily uses reference counting to keep track of how many references exist to an object. Each object in Python has an associated reference count. When the reference count drops to zero, the object is deallocated.\n"," - Cyclic Garbage Collection- One limitation of reference counting is that it cannot handle reference cycles—situations where objects reference each other, preventing their reference counts from reaching zero.\n","\n","16. What is the purpose of the else block in exception handling?\n"," - Purpose of else block are to execute code that should only run when the try block succeeds without exceptions.\n"," -To keep the try block focused on the code that may raise exceptions, while moving the \"success path\" logic to the else block for better readability and clarity.\n"," - To differentiate between code that is responsible for handling errors (in the except block) and code that should run after successful execution of the try block.\n","\n","17. What are the common logging levels in Python?\n"," - Python's logging module defines several logging levels, which represent the severity of a log message. These levels are organized hierarchically, with higher levels indicating more critical events.\n"," - DEBUG- Used for detailed debugging information, typically only needed during development.\n"," - INFO- Used for general informational messages about the program's execution.\n"," - WARNING- Used to indicate potential issues that might require attention but don't necessarily represent errors.\n"," - ERROR- Used to report errors that have occurred during program execution.\n"," - CRITICAL- Used to report critical errors that might require immediate action.\n","\n","18. What is the difference between os.fork() and multiprocessing in Python?\n"," - os.fork()- It is a low-level system call that creates a new process by duplicating the existing process. This new process is called the \"child\" process, while the original process is the \"parent\" process.\n"," - The child process initially shares the same memory space as the parent process. However, any changes made by either process after the fork are not reflected in the other process's memory.\n"," - multiprocessing- The multiprocessing module provides a higher-level interface for creating and managing processes in Python. It abstracts away many of the low-level details of process creation and communication.\n"," - It offers various ways to create processes, including using Process objects, pools of processes, and queues for inter-process communication.\n","\n","19. What is the importance of closing a file in Python?\n"," - Closing a file in Python is crucial for several reasons, as it ensures proper resource management, data integrity, and prevents potential issues related to open files.\n"," - When you open a file using Python's open() function, you create a connection between your program and the file on the disk. Closing the file is necessary to terminate this connection properly.\n"," - Closing the file ensures that any remaining data in the buffer is flushed to the file, guaranteeing that all data is saved properly. If you don't close the file, there is a risk of data loss or corruption.\n","\n","20.  What is the difference between file.read() and file.readline() in Python?\n"," - file.read()- It reads the entire contents of the file as a single string.\n"," - It reads from the current file position until the end of the file. If you specify a size argument, it reads at most that many bytes.\n"," - file.readline()- It reads a single line from the file.\n"," -  It reads from the current file position until it encounters a newline character (\\n) or the end of the file. The newline character is included in the returned string.\n","\n","21. What is the logging module in Python used for?\n"," - It Helps identify and resolve issues in the code by logging detailed information during development.\n"," - Logs warnings, errors, or critical failures in the application for troubleshooting.\n"," - Keeps a record of events to analyze the application's performance and behavior over time.\n"," - Provides a trail of actions or changes, useful for security and compliance purposes.\n","\n","22. What is the os module in Python used for in file handling?\n"," - The os module in Python provides a way to interact with the operating system. It includes functions for file handling, directory management, and other system-level tasks.\n"," - When it comes to file handling, the os module is widely used for performing operations like creating, reading, writing, deleting, or modifying files and directories.\n","\n","23. What are the challenges associated with memory management in Python?\n"," - Memory management in Python is primarily handled automatically by the Python interpreter using reference counting and a garbage collection (GC) system. While this simplifies memory management for developers, it is not entirely free of challenges.\n"," - Below are some common challenges associated with memory management in Python-\n"," - Overhead of Garbage Collection\n"," - Memory Fragmentation\n"," - Cyclic References\n"," - Limited Control\n","\n","24. How do you raise an exception manually in Python?\n"," - You can manually raise exceptions in Python using the raise statement. This allows you to signal errors or exceptional conditions in your code.\n"," - ExceptionType: The type of exception you want to raise (e.g., ValueError, TypeError, RuntimeError).\n"," - Optional error message: A string providing more details about the error.\n","\n","25. Why is it important to use multithreading in certain applications?\n"," - Multithreading is a programming concept where multiple threads (smaller units of a process) run concurrently within a single program. It is essential in certain applications to improve performance, responsiveness, and efficient use of system resources.\n"," - Below are the key reasons why multithreading is important-\n"," - Improves Application Responsiveness\n"," - Concurrency for I/O-Bound Tasks\n"," - Efficient Use of Multi-Core Processors\n"," - Faster Execution of Parallelizable Tasks\n"," - Simplifies Program Design\n"," - Real-Time and Time-Sensitive Applications"],"metadata":{"id":"kmex5FvH7_fC"}},{"cell_type":"markdown","source":["#Practical Questions"],"metadata":{"id":"mIm1ZA7gx_SU"}},{"cell_type":"code","source":["# 1. How can you open a file for writing in Python and write a string to it?\n","def open_file(file_name, text):\n","    try:\n","        with open(file_name, 'w') as file:\n","            file.write(text)\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"riraFOq8yu35","executionInfo":{"status":"ok","timestamp":1736270626244,"user_tz":-330,"elapsed":340,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":1,"outputs":[]},{"cell_type":"code","source":["open_file(\"test.txt\", \"Hello World\")"],"metadata":{"id":"iKD1RexIyvdR","executionInfo":{"status":"ok","timestamp":1736270658218,"user_tz":-330,"elapsed":325,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":2,"outputs":[]},{"cell_type":"code","source":["# 2.Write a Python program to read the contents of a file and print each line.\n","def read_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            for line in file:\n","                print(line.strip())\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"nYSqkrlNy7cw","executionInfo":{"status":"ok","timestamp":1736270740560,"user_tz":-330,"elapsed":344,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":3,"outputs":[]},{"cell_type":"code","source":["read_file(\"test.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"iGtDreaMzPla","executionInfo":{"status":"ok","timestamp":1736270761666,"user_tz":-330,"elapsed":447,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"6aec4206-b8b7-4fa0-afb3-5a1062358d49"},"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["Hello World\n"]}]},{"cell_type":"code","source":["# 3.How would you handle a case where the file doesn't exist while trying to open it for reading?\n","def open_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            for line in file:\n","                print(line.strip())\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"x16Odz6AzUhj","executionInfo":{"status":"ok","timestamp":1736270849983,"user_tz":-330,"elapsed":353,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":5,"outputs":[]},{"cell_type":"code","source":["open_file(\"test1.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"LJ5yI2A5zqP2","executionInfo":{"status":"ok","timestamp":1736270860914,"user_tz":-330,"elapsed":5,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"a60f634f-ec08-4aa6-df6c-837206cf1696"},"execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: File 'test1.txt' not found.\n"]}]},{"cell_type":"code","source":["# 4.  Write a Python script that reads from one file and writes its content to another file.\n","def read_file(source, destination):\n","    try:\n","        with open(source, 'r') as source_file:\n","            with open(destination, 'w') as destination_file:\n","                for line in source_file:\n","                    destination_file.write(line)\n","        print(\"File copied successfully.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"VEdYFRIJzs7M","executionInfo":{"status":"ok","timestamp":1736271083802,"user_tz":-330,"elapsed":310,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":8,"outputs":[]},{"cell_type":"code","source":["read_file(\"test.txt\", \"test2.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PhW5wcOk0jWJ","executionInfo":{"status":"ok","timestamp":1736271096607,"user_tz":-330,"elapsed":367,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"06a145aa-1a80-4ae8-a2d1-d2083a723b59"},"execution_count":9,"outputs":[{"output_type":"stream","name":"stdout","text":["File copied successfully.\n"]}]},{"cell_type":"code","source":["#5. How would you catch and handle division by zero error in Python?\n","def divide_numbers(a, b):\n","    try:\n","        result = a / b\n","        return result\n","    except ZeroDivisionError:\n","        print(\"Error: Division by zero is not allowed.\")\n","        return None"],"metadata":{"id":"q0eyxg160mds","executionInfo":{"status":"ok","timestamp":1736271145516,"user_tz":-330,"elapsed":354,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":10,"outputs":[]},{"cell_type":"code","source":["divide_numbers(10, 0)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"eUUTI2lX0yZu","executionInfo":{"status":"ok","timestamp":1736271188066,"user_tz":-330,"elapsed":368,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"5dd5b0fb-9630-4522-dc3d-4f0265649724"},"execution_count":11,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: Division by zero is not allowed.\n"]}]},{"cell_type":"code","source":["# 6.Write a Python program that logs an error message to a log file when a division by zero exception occur.\n","import logging\n","\n","logging.basicConfig(filename='error.log', level=logging.ERROR)\n","\n","def divide_numbers(a, b):\n","    try:\n","        result = a / b\n","        return result\n","    except ZeroDivisionError:\n","        logging.error(\"Division by zero error occurred.\")\n","        return None"],"metadata":{"id":"b40-11QS08zN","executionInfo":{"status":"ok","timestamp":1736271503052,"user_tz":-330,"elapsed":405,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":15,"outputs":[]},{"cell_type":"code","source":["a= divide_numbers(10, 0)\n","if a is not None:\n","    print(a)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yAtFt9oq1Pmm","executionInfo":{"status":"ok","timestamp":1736271517804,"user_tz":-330,"elapsed":377,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"da54c837-3a5f-4773-fdac-b957ff304530"},"execution_count":16,"outputs":[{"output_type":"stream","name":"stderr","text":["ERROR:root:Division by zero error occurred.\n"]}]},{"cell_type":"code","source":["# 7.How do you log information at different levels (INFO, ERROR, WARNING) in Python using the logging module?\n","import logging\n","\n","logging.basicConfig(level=logging.DEBUG)\n","logging.debug(\"This is a debug message.\")\n","\n","logging.basicConfig(level=logging.INFO)\n","logging.info(\"This is an info message.\")\n","\n","logging.basicConfig(level=logging.ERROR)\n","logging.error(\"This is an error message.\")\n","\n","logging.basicConfig(level=logging.WARNING)\n","logging.warning(\"This is a warning message.\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yXWhS3gm1VsS","executionInfo":{"status":"ok","timestamp":1736273227291,"user_tz":-330,"elapsed":406,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"0837e692-5738-40d8-f945-b0f04268e640"},"execution_count":36,"outputs":[{"output_type":"stream","name":"stderr","text":["ERROR:root:This is an error message.\n","WARNING:root:This is a warning message.\n"]}]},{"cell_type":"code","source":["# 8.Write a program to handle a file opening error using exception handling.\n","def open_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            for line in file:\n","                print(line.strip())\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"bO2Sqedt24o-","executionInfo":{"status":"ok","timestamp":1736272037180,"user_tz":-330,"elapsed":404,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":21,"outputs":[]},{"cell_type":"code","source":["open_file(\"test1.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Y4djsrET3vbh","executionInfo":{"status":"ok","timestamp":1736272070919,"user_tz":-330,"elapsed":353,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"58e624d5-3419-43a5-b7b1-584b541dbe85"},"execution_count":24,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: File 'test1.txt' not found.\n"]}]},{"cell_type":"code","source":["# 9.How can you read a file line by line and store its content in a list in Python?\n","def read_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            lines = file.readlines()\n","            return lines\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")\n","        return None"],"metadata":{"id":"kNjEkfSl4CPt","executionInfo":{"status":"ok","timestamp":1736272136934,"user_tz":-330,"elapsed":532,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":25,"outputs":[]},{"cell_type":"code","source":["read_file(\"test.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"x2hVxZhM4kaV","executionInfo":{"status":"ok","timestamp":1736272148106,"user_tz":-330,"elapsed":6,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"70a75a84-0beb-4492-f9a3-2d10687bd93e"},"execution_count":26,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['Hello World']"]},"metadata":{},"execution_count":26}]},{"cell_type":"code","source":["# 10. How can you append data to an existing file in Python?\n","def append_data(file_name, data):\n","    try:\n","        with open(file_name, 'a') as file:\n","            file.write(data)\n","            print(\"Data appended successfully.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"8VRGyBko4nP8","executionInfo":{"status":"ok","timestamp":1736272294722,"user_tz":-330,"elapsed":601,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":27,"outputs":[]},{"cell_type":"code","source":["append_data(\"test.txt\", \"Hello World\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"JCVp7SXY5K60","executionInfo":{"status":"ok","timestamp":1736272302053,"user_tz":-330,"elapsed":345,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"edf0ba3a-776b-41f8-c92f-0ca2c6e00473"},"execution_count":28,"outputs":[{"output_type":"stream","name":"stdout","text":["Data appended successfully.\n"]}]},{"cell_type":"code","source":["# 11. Write a Python program that uses a try-except block to handle an error when attempting to access a dictionary key that doesn't exist.\n","def access_key(dictionary, key):\n","    try:\n","        value = dictionary[key]\n","        return value\n","    except KeyError:\n","        print(f\"Error: Key '{key}' does not exist in the dictionary.\")\n","        return None"],"metadata":{"id":"qfn00VZW5MxL","executionInfo":{"status":"ok","timestamp":1736272589942,"user_tz":-330,"elapsed":405,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":29,"outputs":[]},{"cell_type":"code","source":["access_key({\"a\": 1, \"b\": 2}, \"c\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"b6Mrabaa6TBd","executionInfo":{"status":"ok","timestamp":1736272604290,"user_tz":-330,"elapsed":418,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"e54f9001-963d-47de-9d00-1a97d9731665"},"execution_count":30,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: Key 'c' does not exist in the dictionary.\n"]}]},{"cell_type":"code","source":["#12. Write a program that demonstrates using multiple except blocks to handle different types of exceptions.\n","def multiple_except():\n","    try:\n","        result = 10 / 0\n","        file = open(\"nonexistent_file.txt\", \"r\")\n","    except ZeroDivisionError:\n","        print(\"Error: Division by zero.\")\n","    except FileNotFoundError:\n","        print(\"Error: File not found.\")\n","    except TypeError:\n","        print(\"Error: Type error.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")\n","    else:\n","        print(\"No exceptions occurred.\")\n","    finally:\n","        print(\"Program execution completed.\")"],"metadata":{"id":"HtFG7n6R6Wh_","executionInfo":{"status":"ok","timestamp":1736272903204,"user_tz":-330,"elapsed":336,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":31,"outputs":[]},{"cell_type":"code","source":["multiple_except()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"lkuik_vd7fh4","executionInfo":{"status":"ok","timestamp":1736272910130,"user_tz":-330,"elapsed":426,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"03896f48-8ca8-4220-d187-f5a7bdfe1e3b"},"execution_count":32,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: Division by zero.\n","Program execution completed.\n"]}]},{"cell_type":"code","source":["# 13. How would you check if a file exists before attempting to read it in Python?\n","def check_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            for line in file:\n","                print(line.strip())\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"MbNecvth7hP4","executionInfo":{"status":"ok","timestamp":1736273070172,"user_tz":-330,"elapsed":5,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":33,"outputs":[]},{"cell_type":"code","source":["check_file(\"test.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fTYhrZsA8INf","executionInfo":{"status":"ok","timestamp":1736273077017,"user_tz":-330,"elapsed":335,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"5fecd4dd-91d1-4439-e96e-14fe117abff2"},"execution_count":34,"outputs":[{"output_type":"stream","name":"stdout","text":["Hello WorldHello World\n"]}]},{"cell_type":"code","source":["# 14. Write a program that uses the logging module to log both informational and error messages.\n","import logging\n","\n","logging.basicConfig(filename='log.txt', level=logging.INFO)\n","\n","logger= logging.getLogger(__name__)\n","def log_info(message):\n","    logging.info(message)\n","    print(\"Info message logged.\")\n","\n","def log_error(message):\n","    logging.error(message)\n","    print(\"Error message logged.\")"],"metadata":{"id":"THO6M4C48J94","executionInfo":{"status":"ok","timestamp":1736273339317,"user_tz":-330,"elapsed":305,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":38,"outputs":[]},{"cell_type":"code","source":["log_info(\"This is an info message.\")\n","log_error(\"This is an error message.\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"aiXrXMON8gme","executionInfo":{"status":"ok","timestamp":1736273363524,"user_tz":-330,"elapsed":338,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"a0fe102a-e77d-4170-9b90-cabdc2c73e94"},"execution_count":39,"outputs":[{"output_type":"stream","name":"stderr","text":["ERROR:root:This is an error message.\n"]},{"output_type":"stream","name":"stdout","text":["Info message logged.\n","Error message logged.\n"]}]},{"cell_type":"code","source":["#15.  Write a Python program that prints the content of a file and handles the case when the file is empty.\n","def print_file_content(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            content = file.read()\n","            if content:\n","                print(content)\n","            else:\n","                print(\"File is empty.\")\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")"],"metadata":{"id":"4P9-4B7c9P7I","executionInfo":{"status":"ok","timestamp":1736273430773,"user_tz":-330,"elapsed":349,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":40,"outputs":[]},{"cell_type":"code","source":["print_file_content(\"test.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"dGi2CDVd9gVJ","executionInfo":{"status":"ok","timestamp":1736273481261,"user_tz":-330,"elapsed":364,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"0568f942-1a88-47f6-e35b-2a9a8fe338d5"},"execution_count":46,"outputs":[{"output_type":"stream","name":"stdout","text":["Hello WorldHello World\n"]}]},{"cell_type":"code","source":["#16. Demonstrate how to use memory profiling to check the memory usage of a small program.\n","def memory_usage():\n","    import psutil\n","    process = psutil.Process()\n","    memory_info = process.memory_info()\n","    return memory_info.rss\n","    print(f\"Memory usage: {memory_info.rss} bytes\")"],"metadata":{"id":"0KM-joh99jDq","executionInfo":{"status":"ok","timestamp":1736273544823,"user_tz":-330,"elapsed":340,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":48,"outputs":[]},{"cell_type":"code","source":["memory_usage()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"e6P8dKHe95u8","executionInfo":{"status":"ok","timestamp":1736273553707,"user_tz":-330,"elapsed":504,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"d12bf9a6-56d9-4fd9-c0ed-d507edc6350a"},"execution_count":49,"outputs":[{"output_type":"execute_result","data":{"text/plain":["128520192"]},"metadata":{},"execution_count":49}]},{"cell_type":"code","source":["#17. Write a Python program to create and write a list of numbers to a file, one number per line.\n","def create_file(file_name, numbers):\n","    try:\n","        with open(file_name, 'w') as file:\n","            for number in numbers:\n","                file.write(str(number) + '\\n')\n","        print(\"File created and numbers written successfully.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"bgbvIi_f9-Tm","executionInfo":{"status":"ok","timestamp":1736273613533,"user_tz":-330,"elapsed":531,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":50,"outputs":[]},{"cell_type":"code","source":["create_file(\"numbers.txt\", [1, 2, 3, 4, 5])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"xrxN5EpQ-M6L","executionInfo":{"status":"ok","timestamp":1736273633887,"user_tz":-330,"elapsed":383,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"afe91c3f-619b-48af-930c-fe33382be3bf"},"execution_count":51,"outputs":[{"output_type":"stream","name":"stdout","text":["File created and numbers written successfully.\n"]}]},{"cell_type":"code","source":["#18. How would you implement a basic logging setup that logs to a file with rotation after 1MB?\n","def logging_setup(log_file):\n","    import logging\n","    from logging.handlers import RotatingFileHandler\n","\n","    logger = logging.getLogger(__name__)\n","    logger.setLevel(logging.INFO)\n","    handler = RotatingFileHandler(log_file, maxBytes=1000000, backupCount=5)\n","    formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')\n","    handler.setFormatter(formatter)\n","    logger.addHandler(handler)\n","    return logger"],"metadata":{"id":"k2gfPCAq-q-V","executionInfo":{"status":"ok","timestamp":1736273738480,"user_tz":-330,"elapsed":309,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":52,"outputs":[]},{"cell_type":"code","source":["logging_setup(\"log.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yp-kkux8-rgA","executionInfo":{"status":"ok","timestamp":1736273765728,"user_tz":-330,"elapsed":319,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"898c72e7-a835-4245-dc04-3e108cfeeae1"},"execution_count":54,"outputs":[{"output_type":"execute_result","data":{"text/plain":["<Logger __main__ (INFO)>"]},"metadata":{},"execution_count":54}]},{"cell_type":"code","source":["# 19. Write a program that handles both IndexError and KeyError using a try-except block.\n","def handle_exceptions(data, index, key):\n","    try:\n","        value = data[index]\n","        print(f\"Value at index {index}: {value}\")\n","    except IndexError:\n","        print(\"Error: Index out of range.\")\n","    except KeyError:\n","        print(\"Error: Key not found in the dictionary.\")"],"metadata":{"id":"8QrCIEtU-vDF","executionInfo":{"status":"ok","timestamp":1736273821174,"user_tz":-330,"elapsed":326,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":56,"outputs":[]},{"cell_type":"code","source":["handle_exceptions({\"a\": 1, \"b\": 2}, 2, \"c\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"53N1YMcp--ZW","executionInfo":{"status":"ok","timestamp":1736273828805,"user_tz":-330,"elapsed":417,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"9550aa53-245d-4f7a-ef32-fd4fb414d18f"},"execution_count":57,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: Key not found in the dictionary.\n"]}]},{"cell_type":"code","source":["#20. How would you open a file and read its contents using a context manager in Python?\n","def open_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            contents = file.read()\n","            print(contents)\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")\n","    except Exception as e:\n","        print(f\"Error: {e}\")"],"metadata":{"id":"HUG7jxwt_BfN","executionInfo":{"status":"ok","timestamp":1736273900890,"user_tz":-330,"elapsed":429,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":58,"outputs":[]},{"cell_type":"code","source":["open_file(\"test.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"lKbxFsm3_TFy","executionInfo":{"status":"ok","timestamp":1736273907901,"user_tz":-330,"elapsed":691,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"9eb7f30e-82a0-4a86-9ba2-7410b2d86699"},"execution_count":59,"outputs":[{"output_type":"stream","name":"stdout","text":["Hello WorldHello World\n"]}]},{"cell_type":"code","source":["#21. Write a Python program that reads a file and prints the number of occurrences of a specific word.\n","def count_occ(file_name, word):\n","    try:\n","        with open(file_name, 'r') as file:\n","            content = file.read()\n","            content = content.lower()\n","            word = word.lower()\n","            occurrences = content.count(word)\n","            return occurrences\n","            print(f\"The word '{word}' appears {occurrences} times in the file.\")\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")"],"metadata":{"id":"9-mccahz_UvN","executionInfo":{"status":"ok","timestamp":1736274067442,"user_tz":-330,"elapsed":386,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":60,"outputs":[]},{"cell_type":"code","source":["count_occ(\"test.txt\", \"hello\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"osntVkVu_7w_","executionInfo":{"status":"ok","timestamp":1736274088199,"user_tz":-330,"elapsed":395,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"b17042f7-1c1c-4d85-94e6-172504f3b4bc"},"execution_count":61,"outputs":[{"output_type":"execute_result","data":{"text/plain":["2"]},"metadata":{},"execution_count":61}]},{"cell_type":"code","source":["#22.How can you check if a file is empty before attempting to read its contents?\n","def check_empty(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            content = file.read()\n","            if content:\n","                print(\"File is not empty.\")\n","            else:\n","                print(\"File is empty.\")\n","    except FileNotFoundError:\n","        print(f\"Error: File '{file_name}' not found.\")"],"metadata":{"id":"r6KPY3bsAA1P","executionInfo":{"status":"ok","timestamp":1736274116597,"user_tz":-330,"elapsed":577,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}}},"execution_count":62,"outputs":[]},{"cell_type":"code","source":["check_empty('test.txt')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2c0DoEMtAHtH","executionInfo":{"status":"ok","timestamp":1736274163857,"user_tz":-330,"elapsed":566,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"d906ad15-56f0-45d6-8ce3-7e5984b3d148"},"execution_count":64,"outputs":[{"output_type":"stream","name":"stdout","text":["File is not empty.\n"]}]},{"cell_type":"code","source":["#23.Write a Python program that writes to a log file when an error occurs during file handling.\n","import logging\n","\n","logging.basicConfig(filename='error.log', level=logging.ERROR)\n","try:\n","    with open('nonexistent_file.txt', 'r') as file:\n","        content = file.read()\n","except FileNotFoundError:\n","    logging.error(\"File not found error occurred.\")\n","    print(\"Error: File not found.\")\n","except Exception as e:\n","    logging.error(f\"An error occurred: {e}\")\n","    print(f\"Error: {e}\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Xln0ZpJkAPOg","executionInfo":{"status":"ok","timestamp":1736274223921,"user_tz":-330,"elapsed":348,"user":{"displayName":"Anup Ray","userId":"11742562675350626298"}},"outputId":"aa9c406a-519f-4d23-cd4c-a481c5e2467f"},"execution_count":65,"outputs":[{"output_type":"stream","name":"stderr","text":["ERROR:root:File not found error occurred.\n"]},{"output_type":"stream","name":"stdout","text":["Error: File not found.\n"]}]},{"cell_type":"code","source":[],"metadata":{"id":"icpNGqG2AhwY"},"execution_count":null,"outputs":[]}]}